LoginClient/LoginClientForm.cs

```diff
@@ -0,0 +1,230 @@
+        private void DisplayRecievedMessage(byte[] message)
+        {
+            Console.WriteLine(Thread.CurrentThread.ManagedThreadId + "in DisplayRecievedMessage");
+            ASCIIEncoding encoder = new ASCIIEncoding();
+            string str = encoder.GetString(message, 0, message.Length);
+
+            if (str == "close")
+            {
+                //MessageLogTB.Text += str + "\r\n";
+                pipeClient.Disconnect();
+                CreateNewPipeClient();
+            }
+            MessageLogTB.Text += str + "\r\n";
+        }
+
+        /// <summary>
+        /// Update the enabled property of the buttons on the form.
+        /// Then resets the focus.
+        /// </summary>
+        private void UpdateForm()
+        {
+            bool connected = pipeClient.IsConnected();
+            bool loggedIn = pipeClient.IsLoggedIn();
+
+            ConnectBtn.Enabled = !connected;
+            DisconnectBtn.Enabled = connected;
+
+            
+            SendBtn.Enabled = loggedIn;
+            SendMessageTB.Enabled = loggedIn;
+            ClearBtn.Enabled = loggedIn;
+
+            if (connected == false)
+            {
+                ConnectBtn.Focus();
+                LoginBtn.Enabled = connected;
+            }
+            else if (loggedIn == false)
+            {
+                AdminUsernameTB.Focus();
+                LoginBtn.Enabled = !loggedIn;
+            }
+            else
+            {
+                SendMessageTB.Focus();
+                LoginBtn.Enabled = !loggedIn;
+            }
+        }
+
+        
+        #region ButtonClickMethods
+
+        /// <summary>
+        /// Attempts to connect to the server.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void ConnectBtn_Click(object sender, EventArgs e)
+        {
+            pipeClient.Connect(PipeNameTB.Text);
+
+            UpdateForm();
+
+        }
+
+        /// <summary>
+        /// Sends the user text to the server and clears the send message box.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void SendBtn_Click(object sender, EventArgs e)
+        {
+            ASCIIEncoding encoder = new ASCIIEncoding();
+
+            if (!String.IsNullOrWhiteSpace(SendMessageTB.Text))
+            {
+                pipeClient.SendMessage(encoder.GetBytes(SendMessageTB.Text));
+                SendMessageTB.Clear();
+            }
+        }
+
+        /// <summary>
+        /// When the disconnect button is clicked the pipeclient is dicconected, 
+        /// the form is updated and a new pipeclient is created.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void DisconnectBtn_Click(object sender, EventArgs e)
+        {
+            pipeClient.Disconnect();
+            CreateNewPipeClient();
```

Comment:
Here, is a potential bug creator.
You disconnect.  And then immediately create a new pipe, instead of shutting everything off and deleting it.

Generally you either.
Create everything on connect.  Kill everything on disconnect.
Or
Have some background pipe service working, 
and connect and disconnect  using it.
But dont kill it until the end of your app.

So this could cause confusion, which can create bugs.